flatnav/distances/InnerProductDistance.h

```diff
@@ -22,36 +24,40 @@ class InnerProductDistance : public DistanceInterface<InnerProductDistance> {
 public:
   InnerProductDistance() = default;
 
-  explicit InnerProductDistance(size_t dim) {
-    _dimension = dim;
-    _data_size_bytes = dim * sizeof(float);
+  explicit InnerProductDistance(size_t dim)
+      : _dimension(dim), _data_size_bytes(dim * sizeof(float)),
+        _distance_computer(std::bind(&InnerProductDistance::defaultDistanceImpl,
+                                     this, std::placeholders::_1,
+                                     std::placeholders::_2,
+                                     std::placeholders::_3)) {
+    setDistanceFunction();
   }
 
   float distanceImpl(const void *x, const void *y,
                      bool asymmetric = false) const {
     (void)asymmetric;
-    // Default implementation of inner product distance, in case we cannot
-    // support the SIMD specializations for special input _dimension sizes.
-    float *p_x = (float *)x;
-    float *p_y = (float *)y;
-    float result = 0;
-    for (size_t i = 0; i < _dimension; i++) {
-      result += p_x[i] * p_y[i];
-    }
-    return 1.0 - result;
+    _distance_computer(x, y, _dimension);
```

Comment:
I really don't like calling distances by function pointer. There's a whole bunch of reasons why (it's an extra indirection, we call the distance function O(10k) times per query, it prevents compiler from optimizing / inlining the distances, it can mess with processor branch prediction, etc). Of course, it's not a foregone conclusion that it'll be much slower in practice - a good, modern optimizing compiler / caching system can probably mitigate the impact so that it's about the same speed. But a static function call will always be at least as fast (possibly faster). 

We've already got an architecture that templates on the distance, using the CRTP to get compile-time polymorphism that avoids call-by-pointer. Can we make SIMD specializations fit into this design? I think this might do it:
1. Write DistanceInterface subclasses for each SIMD specialization that we want to use (compile time).
2. In the Python bindings, define indexes for each specialization (compile time) and then (at runtime) decide which one to use based on dimension, instruction set support, etc.

The only pitfall here is that we wouldn't be able to load a PyIndex<InnerProductDistanceWithSIMD, int> as a PyIndex<InnerProductDistance, int>. To fix that, we can probably have all the InnerProductDistanceSpecializations inherit from InnerProductDistance. WDYT?

Reply:
Note: I don't think we should do this now. This should go into a refactoring PR, due to time pressure for experiments. We could also benchmark to see whether call-by-pointer is that bad (I worry about the loss of inlining in beam search, but maybe it's no big deal on recent compilers / platforms).

Reply:
I agree that it would perhaps be a bit more efficient to not use the function pointer. It's supposed to be resolved once after distance construction so that calls to `distance()` won't go through an extra layer of indirection, so I'm a bit skeptical about by how much it would help to not use the pointer. In any case, I like your suggestion too, and I'll revisit this to see if it performs faster to do it the way you suggested.